### ISSUE TITLE
New definition for colors

### ISSUE THREAD

**Issue by baetheus** [nullpub/css#12]
The color scheming variables and setup is rather shallow. A new definition for colors should be constructed with the following requirements.

Colors should be grouped by theme

A color theme should consist of the following based on interaction types:

* Base Color - bf
* Base Background Color - bb
* Base Border Color - be
* Hover Color - hf
* Hover Background Color - hb
* Hover Border Color - he
* Focus Color - ff
* Focus Background Color - fb
* Focus Border Color - fe
* Disabled Color - df
* Disabled Background Color - db
* Disabled Border Color - de

The required basic themes should be

* Primary - pri
* Secondary - sec
* Progress - prg
* Success - suc
* Warning - wrn
* Error - err
* Link - lnk

Usage classes should be available for setting:

* Complete Theme (With Psuedo Classes) - ct
* Interaction Theme - cba, cha, cfa, cda
* Specific Interaction - c[interaction] (ie. cbf, cbb, cbe for base foreground, base background, base edge)
* Other psuedo classes (:hover, :disabled, :focus) should only be available via custom properties in after-defined classes

Examples

* Primary Theme - ct-pri
* Progress Focus Theme - cfa-prg
* Warning Disabled Foreground Color - cdf-wrn

**Comment by baetheus** [comment]
Other Ideas:

* It would be pretty sweet to have another usage for the focus-within psuedo selector. (ie. ct-sec-fi) Only defined for .ct-[theme]-fi:focus-within (this would need to be declared after ct-[theme])
* Declaration order should be as follows to gracefully allow specificity of replacement to follow:
  * ct-[theme]
  * c[interaction]a-[theme]
  * c[interaction][target]-[theme]
* Background bg should use the background, not the background-color property.
* Border color should be pulled out of the border section and into the color section.
* Themes should require Base settings (Color, Background, Border Color) but should inherit for other interactions if they are undefined. This is to make adding additional simple themes easier.

**Comment by baetheus** [comment]
The scss map should follow this structure:

```scss
$themes: (
  'primary': (
    'class-label': 'pri',
    'var-label': 'pri',
    'base-foreground': #000000,
    'base-background': #ffffff,
    'base-edge': #000000,
    'hover-foreground': #111111,
    'hover-background': #efefef,
    'hover-edge': #111111,
    'focus-foreground': #000000,
    'focus-background': #fffbf5,
    'focus-edge': #000000,
    'disabled-foreground': #222222,
    'disabled-background': #dedede,
    'disabled-edge': #222222,
  ),
)
```

**Comment by baetheus** [comment]
Released simplified theme structure in 9adc12709ec768cfd23bf17c88748211a7a551bd

Instead of having a theme have some of base, hover, focus, disabled sections, instead make themes flexible enough to work with any existing ui system. Thus, a theme represents three properties: color, background, and border-color. The structure is as follows:

```scss
$themes: (
  (
    'label': 'focus',
    'foreground': #555555,
    'background': #ffb700,
    'edge': #555555,
    'psuedo': ('focus', 'focus-within'),
  ),
);
```

Every map property is required. The default builds come with themes for base, primary, secondary, link, visited, focus, required, error, and disabled. But because the theme doesn't require adherence to any particular organization structure, it can be used to extend material-ui, bootstrap, etc. This pattern also cuts about 2K from the gzipped full build.
